plone/app/z3cform/widget.py

```diff
@@ -619,6 +620,34 @@ def _base_args(self):
 
         return args
 
+    def items(self):
+        """Return item for the widget values for the display template
+
+        Query the catalog for the widget-value (uuids) to only display items
+        that the user is allowed to see. Accessing the value with e.g.
+        getattr(self.context, self.__name__) would yield the items unfiltered.
+        To keep the order intact query each uuid individually.
+        Uses IContentListing for easy access to MimeTypeIcon and more.
+        """
+        results = []
+        if not self.value:
+            return results
+        separator = getattr(self, 'separator', ';')
+        uuids = self.value.split(separator)
+        if not isinstance(uuids, (list, tuple, set)):
+            uuids = [uuids]
+
+        try:
+            catalog = getToolByName(self.context, 'portal_catalog')
+        except AttributeError:
+            catalog = getToolByName(getSite(), 'portal_catalog')
+
+        for uuid in uuids:
+            brains = catalog(UID=uuid)
```

Comment:
Isn't this code equivalent to:
```    
results = list(catalog(UID=uids))
```
?

Reply:
No, this is to make sure that the order of item is the same as when editing. 
I now found a better way to do that:
```python
        brains = catalog(UID=uuids)
        # restore original order
        brains = {i.UID: i for i in brains}
        for uuid in uuids:
            brain = brains.get(uuid, None)
            if brain:
                results.append(brain)
        return IContentListing(results)
```

Reply:
`{i.UID: i for i in brains}` will not be sorted in PY2 (we can even say that we do not care, I do not see it as a blocker given that it is already an improvement with respect to before).

Something like this should work as well (untested and maybe too dense):
```python
results = sorted(brains, key=lambda brain: uuids.index(brain.UID))
```
if `uuids` is a list.

Reply:
Nice one. I don't like lambdas but that's a great solution.